src/core/thread/mle.cpp

```diff
@@ -2260,13 +2260,27 @@ otError Mle::HandleAdvertisement(const Message &aMessage, const Ip6::MessageInfo
 
     otLogInfoMle(GetInstance(), "Received advertisement from %04x", sourceAddress.GetRloc16());
 
+    aMessageInfo.GetPeerAddr().ToExtAddress(macAddr);
+
     if (mRole != OT_DEVICE_ROLE_DETACHED)
     {
-        SuccessOrExit(error = netif.GetMle().HandleAdvertisement(aMessage, aMessageInfo));
+        if (mDeviceMode & ModeTlv::kModeFFD)
+        {
+            SuccessOrExit(error = netif.GetMle().HandleAdvertisement(aMessage, aMessageInfo));
+        }
+        else
+        {
+            // Remove stale parent
+            if ((neighbor = GetNeighbor(macAddr)) != NULL &&
+                neighbor->GetRloc16() != sourceAddress.GetRloc16())
```

Comment:
Simplify here by using`if (mParent.GetRloc16() == sourceAddress.GetRloc16())` instead? Since MTDs should have only one neighbor (its parent).

Reply:
The first condition also ensures this advertisement is from its parent.